packages/database/src/cachers/EntityHierarchyCacher.js

```diff
@@ -0,0 +1,120 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2020 Beyond Essential Systems Pty Ltd
+ */
+
+import { AsyncTaskQueue } from '@tupaia/utils';
+import { ORG_UNIT_ENTITY_TYPES } from '../modelClasses/Entity';
+import { TYPES } from '../types';
+const { PROJECT, ENTITY, ENTITY_RELATION, ANCESTOR_DESCENDANT_RELATION } = TYPES;
+const BATCH_SIZE = 5;
+
+export class EntityHierarchyCacher {
+  constructor(database) {
+    this.database = database;
+    this.generationsVisited = new Set();
+  }
+
+  getGenerationKey = (entities, hierarchyId) =>
+    `${entities.map(e => e.code).join('-')}/${hierarchyId}`;
+
+  async buildAndCacheAll() {
+    // projects are the root entities of every full tree, so start with them
+    const projects = await this.database.find(PROJECT);
```

Comment:
I like this approach. Also good to proactively build project caches rather than manually 'clicking' in prjects.